src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,494 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    using static SyntaxFactory;
+
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider
+    {
+        [Flags]
+        private enum NodeKind
+        {
+            Conjunction = 1,
+            VarPattern = 1 << 1,
+            TypePattern = 1 << 2,
+            NotNullPattern = 1 << 3,
+            ConstantPattern = 1 << 4,
+            PatternMatch = 1 << 5,
+            Evaluation = 1 << 6,
+            PositionalPattern = 1 << 7,
+            DiscardPattern = 1 << 8,
+        }
+
+        private abstract class AnalyzedNode
+        {
+            public abstract NodeKind Kind { get; }
+
+            public virtual bool IsReduced { get; private set; }
```

Comment:
This seems wasteful. This reducer *is* stateful, we want to mark certain intersection paths as non-trivial but a single boolean value suffices. We can factor out a partial visitor and track triviality as a field there.